plugins/network-elements/nicira-nvp/src/main/java/com/cloud/network/element/NiciraNvpElement.java

```diff
@@ -239,8 +244,51 @@ public boolean implement(Network network, NetworkOffering offering, DeployDestin
          * multiple operations that should be done only once.
          */
 
-        // Implement SourceNat immediately as we have al the info already
-        if (networkModel.isProviderSupportServiceInNetwork(network.getId(), Service.SourceNat, Provider.NiciraNvp)) {
+        if (network.getGuestType().equals(GuestType.Shared)){
```

Comment:
This block (and the next) add to an already large method. It would be best to refactor it to new methods (or a new object altogether).